apps/app/src/app/shared/layout/page-nav/page-nav.component.ts

```diff
@@ -15,17 +34,42 @@ import { pageNavTmpl } from '~/app/shared/layout/page-nav/page-nav-outlet.compon
 				<div class="space-y-2 px-1">
 					<h3 class="font-medium">On this page</h3>
 					<ul class="m-0 flex list-none flex-col">
-						<ng-content />
+						@for (link of links(); track link.id) {
+							<spartan-page-nav-link [ngClass]="{ 'pl-4': link.isNested }" [fragment]="link.id" [label]="link.label" />
+						} @empty {
+							@if (isDevMode()) {
+								[DEV] Nothing to see here!
+							}
+						}
 					</ul>
 				</div>
 			</hlm-scroll-area>
 		</ng-template>
 	`,
 })
-export class PageNavComponent implements AfterViewInit, OnDestroy {
+export class PageNavComponent implements OnInit, AfterViewInit, OnDestroy {
 	@ViewChild('pageNav', { static: true })
 	pageNavTpl?: TemplateRef<unknown>;
 
+	isDevMode = signal(isDevMode());
+	links = signal<SamePageAnchorLink[]>([]);
+
+	private page: HTMLElement = (inject(ElementRef).nativeElement as HTMLElement).previousSibling as HTMLElement;
+
+	ngOnInit() {
+		const selectors = ['[spartanMainSection] spartan-section-sub-heading', '[spartanMainSection] > h3'];
+		const headings = Array.from(this.page.querySelectorAll(selectors.join(',')));
+		const links = headings.map((element) => {
+			const { id, children, localName, textContent } = element;
+			const isSubHeading = localName === 'spartan-section-sub-heading';
+			const label = (isSubHeading ? children[0].childNodes[0].textContent : textContent) ?? '[DEV] Empty heading!';
+			if (isDevMode() && id === '') {
```

Comment:
Just an idea, what do you think of filtering h3 tags without an id?

Reply:
If we do want to skip that, it's an easy selector to add (`h3:not([id])`), but I don't think we should skip them. Anything that isn't a major heading should be downgraded to `h4` status instead.

My reasoning is that eventually, we would want to switch to Markdown/MDX for authoring pages, and when a ToC is generated from Markdown, all headings show up in the ToC (as allowed). We can opt out of certain tags entirely, like skipping h4, h5, and h6, but supporting only _some_ h3 tags while showing other h3 tags conditionally won't be possible later without a custom component (which takes it out of markdown context and into MDX territory).

The way we add IDs now won't be needed later, as a heading in markdown can't have an `id`, and the markdown parser is responsible for creating the slug and adding it as the ID.

Reply:
These are valid points. I think it is fine and i am looking forward to the markdown implementation. ;)